File: #0688.knight-probability-in-chessboard.cpp
class Solution {
public:
    double knightProbability(int n, int k, int row, int column) {
        
        if (k == 0) {
            return 1.00000;
        }

        vector<vector<vector<double>>>dp(n,
            vector<vector<double>>(n, vector<double>(k + 1, -1)));
        return recursion(n, k, row, column, dp);
    }

private:
    double recursion(int n,int k,int row,int col, vector<vector<vector<double>>>&dp) {

        if (row<0 || row>=n || col>=n || col<0) {
            return 0;
        }

        if (k == 0) {
            return 1;
        }

        if (dp[row][col][k] != -1) {
            return dp[row][col][k];
        }

        int coli[] = {-1, -2, -1, -2, +1, +2, +1, +2};
        int colj[] = {-2, -1, +2, +1, -2, -1, +2, +1};
        double result = 0;
        for (int i = 0; i < 8; i++) {
             result += recursion(n, k - 1, row + coli[i], col + colj[i], dp);
        }
        return dp[row][col][k] = result / (double)8;
    }    
};

File: #0688.knight-probability-in-chessboard.js
/**
 * @param {number} n
 * @param {number} k
 * @param {number} row
 * @param {number} column
 * @return {number}
 */
var knightProbability = function(n, k, row, column) {
    let directions = [
        [-2, -1], [-1, -2], [1, -2],
        [2, -1], [2, 1], [1, 2], 
        [-1, 2], [-2, 1]];
    
    let dp = new Array(k + 1)
        .fill(0)
        .map(() => new Array(n).fill(0)
        .map(() => new Array(n).fill(0)))
    ;
    
    dp[0][row][column] = 1;
    for (let h = 1; h <= k; h++) {
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                for (let [dx, dy] of directions) {
                    let newX = i + dx;
                    let newY = j + dy;
                    
                    if (newX >= 0 && newX < n && newY >=0 && newY < n) {
                        dp[h][i][j] += dp[h - 1][newX][newY] / 8;
                    }
                }
            }
        }
    }

    let result = 0;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n ; j++) {
            result += dp[k][i][j]
        }
    }
    return result;
};
